UnitTestFrameworkPkg/GoogleTest.md

```diff
@@ -0,0 +1,358 @@
+  EFI_STATUS                      Status;
+
+  // Any random magic number can be used for these tests
+  SecureBootMode = 0xAB;
+
+  EXPECT_CALL(RtServicesMock, gRT_SetVariable)
+    .WillOnce(Return(EFI_INVALID_PARAMETER));
+
+  Status = SetSecureBootMode(SecureBootMode);
+  EXPECT_EQ(Status, EFI_INVALID_PARAMETER);
+}
+```
+
+Keep in mind that this test is written to verify that `SetSecureBootMode()` will
+return `EFI_INVALID_PARAMETER` when the call to `gRT->SetVariable()` within the
+implementation of `SetSecureBootMode()` returns `EFI_INVALID_PARAMETER`. With that
+in mind, let's discuss how a mock function is used to accomplish this in the test.
+
+In this test case, the `MockUefiRuntimeServicesTableLib` interface is instantiated as
+`RtServicesMock` which enables its associated mock functions. These interface
+instantiations that contain the mock functions are very important for mock function
+based unit tests because without these instantiations, the mock functions within that
+interface will not exist and can not be used.
+
+The next line of interest is the `EXPECT_CALL`, which is a standard part of the gMock
+framework. This macro is telling the test that a call is expected to occur to a
+specific function on a specific interface. The first argument is the name of the
+interface object that was instantiated in this test, and the second argument is the
+name of the mock function within that interface that is expected to be called. The
+`WillOnce(Return(EFI_INVALID_PARAMETER))` associated with this `EXPECT_CALL` states
+that the `gRT_SetVariable()` function (remember from earlier in this documentation
+that this refers to the `gRT->SetVariable()` function) will be called once during
+this test, and when it does get called, we want it to return `EFI_INVALID_PARAMETER`.
+
+Once this `EXPECT_CALL` has been setup, the call to `SetSecureBootMode()` occurs in
+the test, and its return value is saved in `Status` so that it can be tested. Based
+on the `EXPECT_CALL` that was setup earlier, when `SetSecureBootMode()` internally
+calls `gRT->SetVariable()`, it returns `EFI_INVALID_PARAMETER`. This value should
+then be returned by `SetSecureBootMode()`, and the
+`EXPECT_EQ(Status, EFI_INVALID_PARAMETER)` verifies this is the case.
+
+There is much more that can be done with `EXPECT_CALL` and mock functions, but we
+will leave those details to be explained in the gMock documentation.
+
+Now it was mentioned earlier that this test case is not written exactly like this
+in the test file, and the next section describes how this test is slightly
+refactored to reduce the total amount of code in the entire test suite.
+
+### GoogleTest - A gMock Test Case (refactored)
+
+The sample test case from `SecureBootVariableLibGoogleTest` in the prior section is
+actually written as shown below.
+
+```cpp
+class SetSecureBootModeTest : public Test {
+  protected:
+    MockUefiRuntimeServicesTableLib RtServicesMock;

```

Comment:
http://google.github.io/googletest/reference/mocking.html#StrictMock

Based on some researched by @kuqin12, I think we need to specify strict mock:
```suggestion
    StrictMock<MockUefiRuntimeServicesTableLib> RtServicesMock;

```

Using the StrictMock should cause uninteresting calls to trigger errors instead of silently reporting them and moving on.